java/memory/src/main/java/org/apache/arrow/memory/util/ArrowBufPointer.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.memory.util;
+
+import io.netty.buffer.ArrowBuf;
+
+/**
+ * Pointer to a memory region within an {@link io.netty.buffer.ArrowBuf}.
+ * It will be used as the basis for calculating hash code within a vector, and equality determination.
+ */
+public final class ArrowBufPointer {
+
+  private ArrowBuf buf;
+
+  private int offset;
+
+  private int length;
+
+  /**
+   * The default constructor.
+   */
+  public ArrowBufPointer() {
+
+  }
+
+  /**
+   * Constructs an Arrow buffer pointer.
+   * @param buf the underlying {@link ArrowBuf}, which can be null.
+   * @param offset the start off set of the memory region pointed to.
+   * @param length the length off set of the memory region pointed to.
+   */
+  public ArrowBufPointer(ArrowBuf buf, int offset, int length) {
+    set(buf, offset, length);
+  }
+
+  /**
+   * Sets this pointer.
+   * @param buf the underlying {@link ArrowBuf}, which can be null.
+   * @param offset the start off set of the memory region pointed to.
+   * @param length the length off set of the memory region pointed to.
+   */
+  public void set(ArrowBuf buf, int offset, int length) {
+    this.buf = buf;
+    this.offset = offset;
+    this.length = length;
+  }
+
+  public ArrowBuf getBuf() {
+    return buf;
+  }
+
+  public int getOffset() {
+    return offset;
+  }
+
+  public int getLength() {
+    return length;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    ArrowBufPointer other = (ArrowBufPointer) o;
+    if (buf == null || other.buf == null) {
+      if (buf == null && other.buf == null) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    if (length != other.length) {
+      return false;
+    }
+
+    int index = 0;
+    while (index + 8 <= length) {
```

Comment:
Similar logic already exists in https://github.com/apache/arrow/blob/master/java/vector/src/main/java/org/apache/arrow/vector/util/ByteFunctionHelpers.java#L44
Can we directly use it here?

Reply:
@tianchen92 Thanks for your kind reminder.

This functionality will be used in the new design of the dictionary encoding, and possibly other parts of the code base.
 
The logic in ByteFunctionHelpers is based on static methods. So the scenario that is based on ByteFunctionHelpers can also use ArrowBufPointer, but not vice versa.

Reply:
I'm not sure I understand this.  If ByteFunctionHelpers was moved to this package couldn't it be used here?

Reply:
@emkornfield , to use ByteFunctionHelpers, we should move it to the arrow-memory module. Do you think it is OK?